lib/mauth/client.rb

```diff
@@ -298,48 +337,114 @@ def authentic?(object)
         end
       end
 
-      # raises InauthenticError unless the given object is authentic
+      # raises InauthenticError unless the given object is authentic. Will only
+      # authenticate with v2 if the environment variable AUTHENTICATE_WITH_ONLY_V2
+      # is set. Otherwise will authenticate with only the highest protocol version present
       def authenticate!(object)
-        authentication_present!(object)
-        time_valid!(object)
-        token_valid!(object)
-        signature_valid!(object)
-      rescue MauthNotPresent => e
-        logger.warn "mAuth signature not present on #{object.class}. Exception: #{e.message}"
-        raise
-      rescue InauthenticError => e
-        logger.error "mAuth signature authentication failed for #{object.class}. Exception: #{e.message}"
-        raise
-      rescue UnableToAuthenticateError => e
-        logger.error "Unable to authenticate with MAuth for #{object.class}. Exception: #{e.message}"
-        raise
+        if authenticate_with_only_v2
+          if authentication_present_v1(object)
+            msg = 'This service requires mAuth v2 mcc-authentication header but only v1 x-mws-authentication is present'
+            raise MAuth::MissingV2Error, msg
+          end
+          authentication_present_v2!(object)
+          time_valid_v2!(object)
+          token_valid_v2!(object)
+          signature_valid_v2!(object)
+        else
+          if authentication_present_v2(object)
```

Comment:
Although you have stated you do not want to change the code too much, this client file is way too big with don't know how many modules in modules.
I'd rather move the v2 methods into a different file, it would be easier to read, all in one spot.

Reply:
@jcarres-mdsol would you prefer splitting the v2 methods into one file or breaking up the client file by splitting off some of the modules into their own files, e.g. `authenticators/local_authenticator.rb`, `security_token_cacher.rb` etc.